### PR TITLE
Fold FalseEdges and FalseUnwind into Goto + discussion.

### PR DESCRIPTION
**Pull request by vext01**
As discussed on MM, there's probably no advantage to having `FalseEdges` and `FalseUnwind` distinct from `Goto`.

I've raised this as a draft because there are some other discussion points:

1) In MIR there are a handful of terminators which are effectively "go here or fail" or "go here or do cleanup". Here are some examples:

```
    /// Drop the Place                                                          
    Drop {                                                                      
        location: Place<'tcx>,                                                  
        target: BasicBlock,                                                     
        unwind: Option<BasicBlock>,                                             
    },                                     
...
    /// Jump to the target if the condition has the expected value,             
    /// otherwise panic with a message and a cleanup target.                    
    Assert {                                                                    
        cond: Operand<'tcx>,                                                    
        expected: bool,                                                         
        msg: AssertMessage<'tcx>,                                               
        target: BasicBlock,                                                     
        cleanup: Option<BasicBlock>,                                            
    },             
...                        
    /// A suspend point                                                         
    Yield {                                                                     
        /// The value to return                                                 
        value: Operand<'tcx>,                                                   
        /// Where to resume to                                                  
        resume: BasicBlock,                                                     
        /// Cleanup to be done if the generator is dropped at this suspend point
        drop: Option<BasicBlock>,                                               
    },
```

We might choose to merge these into a regular "if then else" style branch.

2) Also the "SwitchInt" terminator is effectively a N-ary jump table. Do we want this to be our "canonical" branching statement? It might be tempting to break a (e.g.) 3-way jump into two binary decisions, but this would cause the block structure to diverge from that of the MIR (which I think we don't want, right?)...

3) We've said that "bytecode" might not be a good term for our IR. So what should we go for? "TIR" -- Tracing IR? "YIR" Yorick IR? I must admit, I'm not keen on "instruction format". If we can agree on something, I'll udate the terminology in my private branch.

Thoughts?

### CODE REVIEW

**Comment by jacob-hughes** [comment]
**1** 
> In MIR there are a handful of terminators which are effectively "go here or fail" or "go here or do cleanup".

A terminator just represents an edge to some other node in the CFG, so I think that this failure or cleanup code can be modeled as a conditional branch to some `BasicBlock` which either: does the cleanup; or (more likely) makes a call to some intrinsic which does the cleanup.

**2**

>It might be tempting to break a (e.g.) 3-way jump into two binary decisions, but this would cause the block structure to diverge from that of the MIR (which I think we don't want, right?)... 

I'm not sure _why_ it's important that we don't diverge block structure from that of the MIR? Isn't the reason for the new IR to create a more convenient format for analysing traces? If splitting a jump table into binary conditional jumps makes traversing (and performing subsequent optimizations) based on a trace more convenient, then shouldn't we do that? I assume it's not yet clear whether that's the case, and as such, I would tend to prefer a solution which leads towards a simpler IR so that it's easier to go back and fix it later.

Intuitively, I wouldn't have thought diverging from the MIR block structure is necessarily a bad thing. We're not throwing control flow information away by lowering MIR into an IR with a different representation of basic block termination, just rejigging it somewhat. (In fairness, we are already diverging from MIR anyway by converting it to SSA form).

MIR is useful for doing things like liveness, reachability, move, and borrow checking (as MIR passes) but Rust relies on LLVM IR for more traditional compiler optimisations, which is why my intuition is that you want some IR which more closely resembles LLVM IR but with better type information.

**3**. I don't really have any thoughts on this, TIR is quite nice.

**tl;dr**
I guess the important distinction is that when created, we're not thinking in terms of MIR anymore.  Instead our new IR represents the parts of MIR that we're interested in, in a more convenient format, while throwing away the stuff we don't care about (source info etc). Which is why I see it as as counterproductive to think of this simply as "diet MIR", and constrain ourselves too rigidly to MIR's structure without good reason.

Additionally, you've probably already seen it, but you might find the original MIR RFC useful for understanding their motivations: https://github.com/rust-lang/rfcs/blob/master/text/1211-mir.md (Their reason for not using SSA is, primarily, because it's not the preferred format for doing borrow checking)

**Comment by vext01** [comment]
Hi Jake,

> I think that this failure or cleanup code can be modelled as a conditional branch to some `BasicBlock` which either: does the cleanup; or (more likely) makes a call to some intrinsic which does the cleanup.

Yep, that's what I was alluding to. But the question is, would it be useful for our JIT to know that an edge is cleanup or drop?

Since I don't yet know the answer, I think we could ignore the edge types for now. Maybe later we could add a notion of "edge flags" which can serve has hints to the JIT if that info turns out to be useful.

> I'm not sure _why_ it's important that we don't diverge block structure from that of the MIR?

It might not be important. However it is an assumption that we've held for some time. I'm happy to throw it out if there's consensus.

> If splitting a jump table into binary conditional jumps makes traversing (and performing subsequent optimizations) based on a trace more convenient, then shouldn't we do that?

I imagine it's a trade-off. As a human I prefer to view one 5-way conditional branch than four binary branches. On the other hand it might -- as you say -- be easier to work with only binary branches when implementing stuff.

**Comment by vext01** [comment]
Thanks for the RFC link by the way. I had not seen it.

**Comment by ltratt** [comment]
I tend to agree with @jacob-hughes here: we should translate the MIR into whatever makes most sense for us, without worrying about MIR itself. In some places we might strongly follow MIR, in some places we might not. From the JIT's perspective, it doesn't matter either way.

**Comment by vext01** [comment]
OK, so unless there are objections, let's translate N-ary branches into multiple binary branches? This will change the block structure a bit (which we've agreed is OK), and possibly make the CFG a little harder for *human* comprehension (which I think we can live with).

**Comment by vext01** [comment]
As I started making adjustments, I realised that diverging from MIR block structure is slightly clumsy. Under the current way that ykrustc works, trace locations are reported as a MIR basic block index. To get the corresponding TIR, we'd need to carry a mapping from MIR block index to TIR block index with us to runtime.

It *might* be possible to generate the TIR earlier in the compiler pipeline and have the trace recorded directly in terms of TIR block indices, but this would require an overhaul.

So our options:
 * Stick with what we have and carry a mapping to runtime.
 * Preserve the MIR block structure, so a TIR and MIR block indices are the ID mapping.
 * Try and munge TIR generation into the trace call insertion MIR pass.

It depends how strongly people feel about moving away from MIR.

**Comment by ltratt** [comment]
Let's do whatever's easiest (in a global sense).

**Comment by vext01** [comment]
@jacob-hughes might not like this, but what's easiest for me right now is to keep the block structure the same as for the MIR. This would mean keeping `SwitchInt` jump tables.

[It also means I can re-use the pre-existing predecessor/successor/dominator helper functions in rustc for SSA generation, but that's somewhat superficial].

**Comment by jacob-hughes** [comment]
That sounds good to me. Seems like a sensible reason to stick to the block structure, I was just questioning doing it for the sake of it.

**Comment by vext01** [comment]
No worries. That's why we have these conversations :)

TBH, I had no good reason either until I sat down to implement it!

**Comment by vext01** [comment]
So we might as well merge this then.
